From 955d6278363474a3e91aaff4b2ef846b094422ca Mon Sep 17 00:00:00 2001
From: xambroz <723625+xambroz@users.noreply.github.com>
Date: Sat, 20 Mar 2021 21:26:59 +0100
Subject: [PATCH] Unhandled result of the pipe function

Added handling of the failed call to pipe function.

Related to https://github.com/radareorg/r2cutter/issues/10

[ 58%] Building CXX object CMakeFiles/r2cutter.dir/widgets/ProcessesWidget.cpp.o
/usr/lib64/ccache/g++ -DCUTTER_ENABLE_GRAPHVIZ -DCUTTER_ENABLE_KSYNTAXHIGHLIGHTING -DCUTTER_SOURCE_BUILD -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_NO_DEBUG -DQT_SVG_LIB -DQT_WIDGETS_LIB -Dr2cutter_EXPORTS -I/home/mambroz/rpmbuild/BUILD/r2cutter-0.1.1/x86_64-redhat-linux-gnu -I/home/mambroz/rpmbuild/BUILD/r2cutter-0.1.1/src -I/home/mambroz/rpmbuild/BUILD/r2cutter-0.1.1/x86_64-redhat-linux-gnu/r2cutter_autogen/include -I/home/mambroz/rpmbuild/BUILD/r2cutter-0.1.1/src/core -I/home/mambroz/rpmbuild/BUILD/r2cutter-0.1.1/src/widgets -I/home/mambroz/rpmbuild/BUILD/r2cutter-0.1.1/src/common -I/home/mambroz/rpmbuild/BUILD/r2cutter-0.1.1/src/plugins -I/home/mambroz/rpmbuild/BUILD/r2cutter-0.1.1/src/menus -I/home/mambroz/rpmbuild/BUILD/r2cutter-0.1.1/src/. -isystem /usr/include/graphviz -isystem /usr/include/qt5 -isystem /usr/include/qt5/QtCore -isystem /usr/lib64/qt5/mkspecs/linux-g++ -isystem /usr/include/qt5/QtWidgets -isystem /usr/include/qt5/QtGui -isystem /usr/include/qt5/QtSvg -isystem /usr/include/qt5/QtNetwork -isystem /usr/include/libr -isystem /usr/include/libr/sdb -isystem /usr/include/capstone -isystem /usr/include/KF5/KSyntaxHighlighting -isystem /usr/include/KF5 -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fvisibility=hidden   -Wall -Wextra -fPIC -std=gnu++11 -o CMakeFiles/r2cutter.dir/widgets/ProcessesWidget.cpp.o -c /home/mambroz/rpmbuild/BUILD/r2cutter-0.1.1/src/widgets/ProcessesWidget.cpp
/home/mambroz/rpmbuild/BUILD/r2cutter-0.1.1/src/widgets/ConsoleWidget.cpp: In member function 'void ConsoleWidget::redirectOutput()':
/home/mambroz/rpmbuild/BUILD/r2cutter-0.1.1/src/widgets/ConsoleWidget.cpp:468:9: warning: ignoring return value of 'int pipe(int*)' declared with attribute 'warn_unused_result' [-Wunused-result]
  468 |     pipe(redirectPipeFds);
      |     ~~~~^~~~~~~~~~~~~~~~~
[ 60%] Building CXX object CMakeFiles/r2cutter.dir/dialogs/MapFileDialog.cpp.o
[ 60%] Building CXX object CMakeFiles/r2cutter.dir/widgets/BacktraceWidget.cpp.o
[ 60%] Building CXX object CMakeFiles/r2cutter.dir/common/CommandTask.cpp.o
---
 src/widgets/ConsoleWidget.cpp | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/widgets/ConsoleWidget.cpp b/src/widgets/ConsoleWidget.cpp
index cb920e29..544a02bb 100644
--- a/src/widgets/ConsoleWidget.cpp
+++ b/src/widgets/ConsoleWidget.cpp
@@ -465,7 +465,10 @@ void ConsoleWidget::redirectOutput()
 
     pipeSocket->connectToServer(pipeName, QIODevice::ReadOnly);
 #else
-    pipe(redirectPipeFds);
+    if (0 > pipe(redirectPipeFds)) {
+        addOutput("Failed to create pipe.");
+        return;       
+    }
     stdinFifoPath = QString(STDIN_PIPE_NAME).arg(QDir::tempPath(), QUuid::createUuid().toString());
     mkfifo(stdinFifoPath.toStdString().c_str(), (mode_t) 0777);
     stdinFile = open(stdinFifoPath.toStdString().c_str(), O_RDWR | O_ASYNC);
